File: assets/js/react/utils/linkify.test.ts
import { describe, expect, it } from "vitest"
import { isSafeUrl, normalizeUrl, parseLinkifiedText } from "./linkify"

describe("normalizeUrl", () => {
  it("adds https to scheme-less domains", () => {
    expect(normalizeUrl("example.com")).toBe("https://example.com")
    expect(normalizeUrl("www.example.com/path")).toBe("https://www.example.com/path")
  })

  it("adds mailto to bare email addresses", () => {
    expect(normalizeUrl("team@example.com")).toBe("mailto:team@example.com")
  })

  it("leaves explicit schemes unchanged", () => {
    expect(normalizeUrl("https://example.com")).toBe("https://example.com")
    expect(normalizeUrl("http://example.com")).toBe("http://example.com")
    expect(normalizeUrl("mailto:a@b.com")).toBe("mailto:a@b.com")
  })
})

describe("isSafeUrl", () => {
  it("allows http and https URLs", () => {
    expect(isSafeUrl("https://example.com")).toBe(true)
    expect(isSafeUrl("http://example.com/path")).toBe(true)
  })

  it("allows scheme-less domains", () => {
    expect(isSafeUrl("example.com")).toBe(true)
    expect(isSafeUrl("www.example.com/about")).toBe(true)
  })

  it("allows bare email addresses", () => {
    expect(isSafeUrl("team@example.com")).toBe(true)
  })

  it("allows mailto URLs", () => {
    expect(isSafeUrl("mailto:user@example.com")).toBe(true)
  })

  it("rejects disallowed schemes", () => {
    expect(isSafeUrl("javascript:alert(1)")).toBe(false)
    expect(isSafeUrl("data:text/html,hello")).toBe(false)
    expect(isSafeUrl("ftp://example.com")).toBe(false)
    expect(isSafeUrl("file:///etc/passwd")).toBe(false)
  })

  it("rejects malformed mailto URLs", () => {
    expect(isSafeUrl("mailto:javascript:alert(1)")).toBe(false)
    expect(isSafeUrl("mailto:")).toBe(false)
    expect(isSafeUrl("mailto:?subject=Hello")).toBe(false)
  })

  it("allows mailto with optional query params", () => {
    expect(isSafeUrl("mailto:team@example.com?subject=Hello")).toBe(true)
  })
})

describe("parseLinkifiedText", () => {
  it("parses markdown links", () => {
    expect(parseLinkifiedText("Visit [site](https://example.com) today.")).toEqual([
      { kind: "text", value: "Visit " },
      { kind: "link", label: "site", href: "https://example.com" },
      { kind: "text", value: " today." },
    ])
  })

  it("parses markdown links without a scheme", () => {
    expect(parseLinkifiedText("Visit [site](example.com) today.")).toEqual([
      { kind: "text", value: "Visit " },
      { kind: "link", label: "site", href: "https://example.com" },
      { kind: "text", value: " today." },
    ])
  })

  it("autolinks bare URLs", () => {
    expect(parseLinkifiedText("See https://example.com for details.")).toEqual([
      { kind: "text", value: "See " },
      { kind: "link", label: "https://example.com", href: "https://example.com" },
      { kind: "text", value: " for details." },
    ])
  })

  it("autolinks scheme-less domains", () => {
    expect(parseLinkifiedText("See example.com for details.")).toEqual([
      { kind: "text", value: "See " },
      { kind: "link", label: "example.com", href: "https://example.com" },
      { kind: "text", value: " for details." },
    ])
  })

  it("autolinks bare mailto URLs", () => {
    expect(parseLinkifiedText("Email mailto:team@example.com today.")).toEqual([
      { kind: "text", value: "Email " },
      { kind: "link", label: "mailto:team@example.com", href: "mailto:team@example.com" },
      { kind: "text", value: " today." },
    ])
  })

  it("autolinks bare email addresses", () => {
    expect(parseLinkifiedText("Contact team@example.com for info.")).toEqual([
      { kind: "text", value: "Contact " },
      { kind: "link", label: "team@example.com", href: "mailto:team@example.com" },
      { kind: "text", value: " for info." },
    ])
  })

  it("parses markdown email links", () => {
    expect(parseLinkifiedText("[Email us](team@example.com)")).toEqual([
      { kind: "link", label: "Email us", href: "mailto:team@example.com" },
    ])
  })

  it("strips trailing punctuation from bare URLs", () => {
    expect(parseLinkifiedText("Go to example.com.")).toEqual([
      { kind: "text", value: "Go to " },
      { kind: "link", label: "example.com", href: "https://example.com" },
      { kind: "text", value: "." },
    ])
  })

  it("leaves unsafe markdown URLs as plain text", () => {
    expect(parseLinkifiedText("[click](javascript:alert(1))")).toEqual([
      { kind: "text", value: "[click](javascript:alert(1))" },
    ])
  })

  it("leaves unsafe bare URLs as plain text", () => {
    expect(parseLinkifiedText("javascript:alert(1)")).toEqual([
      { kind: "text", value: "javascript:alert(1)" },
    ])
  })

  it("leaves malformed mailto as plain text", () => {
    expect(parseLinkifiedText("mailto:javascript:alert(1)")).toEqual([
      { kind: "text", value: "mailto:javascript:alert(1)" },
    ])
  })

  it("does not autolink domains after disallowed schemes", () => {
    for (const input of ["[site](ftp://example.com)", "ftp://example.com"]) {
      const segments = parseLinkifiedText(input)
      expect(segments.every((s) => s.kind === "text")).toBe(true)
      expect(segments.map((s) => (s.kind === "text" ? s.value : "")).join("")).toBe(input)
    }
  })

  it("returns plain text when there are no links", () => {
    expect(parseLinkifiedText("Just a description.")).toEqual([
      { kind: "text", value: "Just a description." },
    ])
  })
})

File: test/storymap/linkify_test.exs
defmodule Storymap.LinkifyTest do
  use ExUnit.Case, async: true

  alias Storymap.Linkify

  describe "render/1" do
    test "returns empty string for nil and blank input" do
      assert Linkify.render(nil) == ""
      assert Linkify.render("") == ""
    end

    test "renders markdown links" do
      html =
        Linkify.render("Visit [site](https://example.com) today.")
        |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(href="https://example.com")
      assert html =~ ">site<"
      assert html =~ "Visit "
      assert html =~ " today."
    end

    test "autolinks bare URLs" do
      html =
        Linkify.render("See https://example.com for details.") |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(href="https://example.com")
      assert html =~ ">https://example.com<"
      assert html =~ "See "
      assert html =~ " for details."
    end

    test "renders markdown links without a scheme" do
      html =
        Linkify.render("Visit [site](example.com) today.")
        |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(href="https://example.com")
      assert html =~ ">site<"
    end

    test "autolinks scheme-less domains" do
      html = Linkify.render("See example.com for details.") |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(href="https://example.com")
      assert html =~ ">example.com<"
      assert html =~ "See "
      assert html =~ " for details."
    end

    test "autolinks bare mailto URLs" do
      html =
        Linkify.render("Email mailto:team@example.com today.") |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(href="mailto:team@example.com")
      refute html =~ ~s(href="https://example.com")
    end

    test "autolinks bare email addresses" do
      html = Linkify.render("Contact team@example.com for info.") |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(href="mailto:team@example.com")
      assert html =~ ">team@example.com<"
      refute html =~ ~s(href="https://example.com")
    end

    test "renders markdown email links" do
      html = Linkify.render("[Email us](team@example.com)") |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(href="mailto:team@example.com")
      assert html =~ ">Email us<"
      refute html =~ ~s(target="_blank")
    end

    test "strips trailing punctuation from bare URLs" do
      html = Linkify.render("Go to example.com.") |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(href="https://example.com")
      assert String.ends_with?(html, ".")
      refute html =~ ~s(href="https://example.com.")
    end

    test "does not link unsafe markdown URLs" do
      html = Linkify.render("[click](javascript:alert(1))") |> Phoenix.HTML.safe_to_string()

      refute html =~ "<a"
      assert html =~ "[click](javascript:alert(1))"
    end

    test "does not link unsafe bare URLs" do
      html = Linkify.render("javascript:alert(1)") |> Phoenix.HTML.safe_to_string()

      refute html =~ "<a"
      assert html =~ "javascript:alert(1)"
    end

    test "does not link malformed mailto URLs" do
      html = Linkify.render("mailto:javascript:alert(1)") |> Phoenix.HTML.safe_to_string()

      refute html =~ "<a"
      assert html =~ "mailto:javascript:alert(1)"
    end

    test "does not link disallowed schemes in markdown" do
      html = Linkify.render("[site](ftp://example.com)") |> Phoenix.HTML.safe_to_string()

      refute html =~ "<a"
      assert html =~ "[site](ftp://example.com)"
    end

    test "allows mailto with query params" do
      html =
        Linkify.render("[Email](mailto:team@example.com?subject=Hello)")
        |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(href="mailto:team@example.com?subject=Hello")
    end

    test "escapes HTML in plain text" do
      html = Linkify.render("<script>alert(1)</script>") |> Phoenix.HTML.safe_to_string()

      refute html =~ "<script>"
      assert html =~ "&lt;script&gt;"
    end

    test "link tags include security attributes" do
      html = Linkify.render("[site](https://example.com)") |> Phoenix.HTML.safe_to_string()

      assert html =~ ~s(target="_blank")
      assert html =~ ~s(rel="noopener noreferrer")
    end
  end
end
